File: 56 merge intervals/MergeIntervals.py
class MergeIntervals(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        if not intervals:
            return []
        res = []
        start = []
        end = []
        n = len(intervals)
        for i in range(n):
            start.append(intervals[i][0])
            end.append(intervals[i][1])
        
        start.sort()
        end.sort()
        s = 0
        i = 0
        while i < n - 1:
            if end[i] >= start[i + 1]:
                i += 1
            else:
                res.append([start[s], end[i]])
                i += 1
                s = i
        res.append([start[s], end[i]])
        
        return res
        

File: 56 merge intervals/MergeIntervals.java
public class MergeIntervals {
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length == 0 || intervals[0].length == 0) {
            return new int[0][0];
        }
        int[] start = new int[intervals.length];
        int[] end = new int[intervals.length];
        for (int i = 0; i < intervals.length; i++) {
            start[i] = intervals[i][0];
            end[i] = intervals[i][1];
        }
        Arrays.sort(start);
        Arrays.sort(end);
        int s = 0;
        int e = 0;
        List<int[]> res = new ArrayList<>();
        while (e < end.length) {
            if (e < end.length - 1 && end[e] >= start[e + 1]) {
                e++;
            } else {
                res.add(new int[]{start[s], end[e]});
                e++;
                s = e;
            }
        }
        int[][] ret = new int[res.size()][2];
        int i = 0;
        for (int[] interval : res) {
            ret[i][1] = interval[1];
            ret[i][0] = interval[0];
            i++;
        }
        return ret;
    }
}